api/read/read.go

```diff
@@ -29,3 +32,49 @@ func ProtoJSON(r io.Reader, m proto.Message) error {
 	}
 	return protojson.Unmarshal(data, m)
 }
+
+// ProtoJSONWithCheck reads JSON from the request body and stores it in the value
+// pointed to by v. Returns false if an error was written; true if not.
+func ProtoJSONWithCheck(w http.ResponseWriter, r io.Reader, m proto.Message) bool {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		var wrapper = struct {
+			Status  int    `json:"code"`
+			Message string `json:"message"`
+		}{
+			Status:  http.StatusBadRequest,
+			Message: err.Error(),
+		}
+		errData, err := json.Marshal(wrapper)
+		if err != nil {
+			panic(err)
+		}
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write(errData)
+		return false
+	}
+	if err := protojson.Unmarshal(data, m); err != nil {
+		if errors.Is(err, proto.Error) {
+			var wrapper = struct {
+				Message string `json:"message"`
+			}{
+				Message: err.Error(),
+			}
+			errData, err := json.Marshal(wrapper)
+			if err != nil {
+				panic(err)
+			}
+			w.Header().Set("Content-Type", "application/json")
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write(errData)
+			return false
+		}
+
+		// fallback to the default error writer
+		render.Error(w, err)
+		return false
+	}
+
+	return true
+}
```

Comment:
The goal of `ProtoJSONWithCheck` was to be able to provide a clearer error message to the user when the input data cannot be unmarshaled. An example error is shown below:

```json
{
    "message": "proto: syntax error (line 7:13): unexpected token \"emails\""
}
```

Without this, it would be a rather generic error without a hint to the user.

Reply:
I think the shortest path to victory here is for @hslatman to expand the functionality of `read.ProtoJSON` so that it bubbles up errors which implement the `render.RenderableError` interface when it encounters `proto.Error`s.

Reply:
Fixed this in def9438ad62e2c71975e2ac1a4f96b6f5ec663a5 followed by 6532c933030e33e46c28ddc5018d17f3d3926720.